File: C++/142-linked-list-cycle-ii.cpp
#include "header.h"

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head, * slow = head;
        do {
            if (fast == nullptr || fast->next == nullptr) return NULL;
            fast = fast->next->next;
            slow = slow->next;
        } while (fast != slow);
        slow = head;
        while (fast != slow) {
            fast = fast->next;
            slow = slow->next;
        }
        return fast;
    }
};

File: JS/142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function (head) {
  let lo = head,
    hi = head;
  do {
    if (!hi?.next?.next) return null;
    lo = lo.next;
    hi = hi.next.next;
  } while (lo != hi);
  lo = head;
  while (lo != hi) {
    lo = lo.next;
    hi = hi.next;
  }
  return hi;
};
